src/configuration.js

```diff
@@ -0,0 +1,252 @@
+     * @return {string | number | boolean}
+     */
+    get(key, defaultValue) {
+        // prefer env vars
+        const envKey = Configuration.ENV_MAP_REVERSED[key] ?? key;
+        const envValue = process.env[envKey] ?? process.env[ENV_VARS[envKey] ?? process.env[`APIFY_${envKey}`]];
+
+        if (envValue != null) {
+            return this._cast(key, envValue, true);
+        }
+
+        // check instance level options
+        if (key in this.options) {
+            return this.options[key];
+        }
+
+        // fallback to defaults
+        return defaultValue ?? Configuration.DEFAULTS[key] ?? envValue;
+    }
+
+    /**
+     * @param {string} key
+     * @param {number | string | boolean} value
+     * @param {boolean} [env=false]
+     * @return {boolean}
+     * @private
+     */
+    _cast(key, value, env = false) {
+        if (key in Configuration.INTEGER_VARS) {
+            return +value;
+        }
+
+        if (key in Configuration.BOOLEAN_VARS) {
+            if (env) {
+                // 0, false and empty string are considered falsy values
+                return !['0', 'false', ''].includes(value.toLowerCase());
+            }
+
+            return !!value;
+        }
+
+        return value;
+    }
+
+    /**
+     * Sets value for given option. Only affects this `Configuration` instance, the value will not be propagated down to the env var.
+     * To reset a value, we can omit the `value` argument or pass `undefined` there.
+     *
+     * @param {string} key
+     * @param {string | number | boolean} [value]
+     */
+    set(key, value) {
+        this.options[key] = value;
+    }
+
+    /**
+     * Creates an instance of ApifyClient using options as defined in the environment variables or in this `Configuration` instance.
+     *
+     * @param {object} [options]
+     * @param {string} [options.token]
+     * @param {string} [options.maxRetries]
+     * @param {string} [options.minDelayBetweenRetriesMillis]
+     * @return {ApifyClient}
+     * @internal
+     */
+    createClient(options = {}) {
+        return new ApifyClient({
+            baseUrl: this.get('apiBaseUrl'),
+            token: this.get('token'),
+            ...options, // allow overriding the instance configuration
+        });
+    }
+
+    /**
+     * Creates an instance of ApifyStorageLocal using options as defined in the environment variables or in this `Configuration` instance.
+     *
+     * @param {object} [options]
+     * @param {string} [options.storageDir]
+     * @param {boolean} [options.enableWalMode=true]
+     * @return {ApifyStorageLocal}
+     * @internal
+     */
+    createLocalStorage(options = {}) {
+        const storageDir = options.storageDir ?? this.get('localStorageDir');
+        const enableWalMode = options.enableWalMode ?? this.get('localStorageEnableWalMode');
+        const storage = new ApifyStorageLocal({ ...options, storageDir, enableWalMode });
```

Comment:
Same here. I'm not quite sure `ApifyStorageLocal` supports existence of multiple instances. Some counts or DB connections might be off. But maybe it will be fine, I haven't really researched it.

Reply:
Oh and it should be `createStorageLocal` not LocalStorage. We flipped it to make it obvious that it's not the Local Storage everyone knows, or something with similar API. Also the product is Apify Storage. So it's Apify Storage local emulation.

Reply:
> I'm not quite sure ApifyStorageLocal supports existence of multiple instances. 

Maybe it does not now, but it should in future. This whole global state madness needs to go, hope we can agree on that :]

Reply:
Oh yeah, sure, but before we merge this, we need to make sure we don't break something. So either work with singleton instances for now, or doublecheck storage local and somehow make sure the AutoscaledPool still works.

Reply:
Both client and storage instances are now cached, based on the parameters (storage dir, api url and token). See 489ff5834e1892202d97e4bf98e46ae6168daf9b for the implementation.